src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/InstrumentProcess.h

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef USER_SPACE_INSTRUMENTATION_INSTRUMENT_PROCESS_H_
+#define USER_SPACE_INSTRUMENTATION_INSTRUMENT_PROCESS_H_
+
+#include <absl/container/flat_hash_set.h>
+
+#include <cstdint>
+
+#include "OrbitBase/Result.h"
+#include "capture.pb.h"
+
+namespace orbit_user_space_instrumentation {
+
+// On the first call to this function we inject OrbitUserSpaceInstrumentation.so into the target
+// process and create the return trampoline. On each call we create trampolines for functions that
+// were not instrumented before and instrument all functions by overwriting the prolog with a jump
+// into the trampoline.
+// Returns the function_id's of the instrumented functions. Note that there is no garantee that we
+// can instrument all the functions in a binary.
+[[nodiscard]] ErrorMessageOr<absl::flat_hash_set<uint64_t>> InstrumentProcess(
+    const orbit_grpc_protos::CaptureOptions& capture_options);
+
+// Undo the instrumentation of the functions. Leaves the library and trampolines in the target
+// process intact. We merely restore the function prologs that were overwritten
+[[nodiscard]] ErrorMessageOr<void> UninstrumentProcess(
+    const orbit_grpc_protos::CaptureOptions& capture_options);
+
+}  // namespace orbit_user_space_instrumentation
```

Comment:
To avoid creating a singleton these functions could be incapsulated in the class.

You can ensure that only one instance of the class is create via CHECK in the factory method, without having the ownership of the class be in the global scope.